core/ordersync/ordersync.go

```
@@ -0,0 +1,507 @@
+				Metadata:     nil,
+			}
+		} else {
+			encodedMetadata, err := json.Marshal(nextReq.Metadata)
+			if err != nil {
+				return err
+			}
+			rawReq = &rawRequest{
+				Type:         TypeRequest,
+				Subprotocols: []string{selectedSubprotocol.Name()},
+				Metadata:     encodedMetadata,
+			}
+		}
+
+		if err := json.NewEncoder(stream).Encode(rawReq); err != nil {
+			s.handlePeerScoreEvent(providerID, psUnexpectedDisconnect)
+			return err
+		}
+
+		rawRes, err := waitForResponse(ctx, stream)
+		if err != nil {
+			return err
+		}
+		s.handlePeerScoreEvent(providerID, psValidMessage)
+
+		subprotocol, found := s.subprotocols[rawRes.Subprotocol]
+		if !found {
+			s.handlePeerScoreEvent(providerID, psSubprotocolNegotiationFailed)
+			return fmt.Errorf("unsupported subprotocol: %s", subprotocol)
+		}
+		selectedSubprotocol = subprotocol
+		res, err := parseResponseWithSubprotocol(subprotocol, providerID, rawRes)
+		if err != nil {
+			s.handlePeerScoreEvent(providerID, psInvalidMessage)
+			return err
+		}
+
+		nextReq, err = subprotocol.HandleOrderSyncResponse(ctx, res)
+		if err != nil {
+			return err
+		}
+		s.handlePeerScoreEvent(providerID, receivedOrders)
+
+		if rawRes.Complete {
+			return nil
+		}
+	}
+}
+
+// shufflePeers randomizes the order of the given list of peers.
+func shufflePeers(peers []peer.ID) {
+	rand.Seed(time.Now().UnixNano())
+	rand.Shuffle(len(peers), func(i, j int) { peers[i], peers[j] = peers[j], peers[i] })
+}
+
+func waitForRequest(parentCtx context.Context, stream network.Stream) (*rawRequest, error) {
+	ctx, cancel := context.WithTimeout(parentCtx, requestResponseTimeout)
+	defer cancel()
+	reqChan := make(chan *rawRequest, 1)
+	go func() {
+		var rawReq rawRequest
+		if err := json.NewDecoder(stream).Decode(&rawReq); err != nil {
+			log.WithFields(log.Fields{
+				"error":     err.Error(),
+				"requester": stream.Conn().RemotePeer().Pretty(),
+			}).Warn("could not encode ordersync request")
```

Doesn't `waitForRequest` want to return this error instead of sending it in the request channel? What am I missing?

---

Good catch. We should be returning the error. Just fixed it.